packages/xo-server/src/api/vm.js

```diff
@@ -1159,12 +1187,77 @@ resume.resolve = {
 
 // -------------------------------------------------------------------
 
-export async function revert({ snapshot }) {
+export const revert = defer(async function($defer, { snapshot }) {
   await this.checkPermissions(this.user.id, [
     [snapshot.$snapshot_of, 'operate'],
   ])
+  const vm = this.getObject(snapshot.$snapshot_of)
+  const { resourceSet } = vm
+  if (resourceSet !== undefined) {
+    const vmUsage = await this.computeVmResourcesUsage(vm)
+    await this.releaseLimitsInResourceSet(vmUsage, resourceSet)
+    $defer.onFailure(() =>
+      this.allocateLimitsInResourceSet(vmUsage, resourceSet, true)
+    )
+
+    // Deallocate IP addresses
+    const vmIpsByVif = {}
+    vm.VIFs.forEach(vifId => {
+      const vif = this.getObject[vifId]
+      vmIpsByVif[vifId] = [
+        ...vif.allowedIpv4Addresses,
+        ...vif.allowedIpv6Addresses,
+      ]
+    })
+    await Promise.all(
+      Object.entries(vmIpsByVif).map(([vifId, ips]) =>
+        this.allocIpAddresses(vifId, null, ips)
+      )
+    )
+    $defer.onFailure(() =>
+      Promise.all(
+        Object.entries(vmIpsByVif).map(([vifId, ips]) =>
+          this.allocIpAddresses(vifId, ips)
+        )
+      )
+    )
+    if (snapshot.resourceSet !== undefined) {
```

Comment:
If the snapshot isn't in the resource set, the new `VM` resource usage will not be taking into account.